File: src/test/java/tool/cmnclslib/cls/UnitTest_ClsCrypt.java
package tool.cmnclslib.cls;

import java.nio.charset.StandardCharsets;
import org.junit.Test;
import static org.junit.Assert.*;
import tool.cmnclslib.mdl.MdlConst;

/**
 * ClsCrypt の単体テストクラスです。
 */
public class UnitTest_ClsCrypt {

    private ClsCrypt crypt = new ClsCrypt();

    @Test
    public void testMd5EncryptDecrypt() {
        crypt.setHashAlgorithm("MD5");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(0);
        assertTrue(crypt.encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, crypt.getResult()));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha1EncryptDecrypt() {
        crypt.setHashAlgorithm("SHA1");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, crypt.getResult()));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha256Size128EncryptDecrypt() {
        crypt.setHashAlgorithm("SHA256");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, crypt.getResult()));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha256Size256EncryptDecrypt() {
        crypt.setHashAlgorithm("SHA256");
        crypt.setKeySize(256);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, crypt.getResult()));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha512Size128EncryptDecrypt() {
        crypt.setHashAlgorithm("SHA512");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, crypt.getResult()));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha512Size256EncryptDecrypt() {
        crypt.setHashAlgorithm("SHA512");
        crypt.setKeySize(256);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, crypt.getResult()));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testMd5DecryptOpenSsl() {
        crypt.setHashAlgorithm("MD5");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(0);
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_MD5_S128_B128_C0));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha1DecryptOpenSsl() {
        crypt.setHashAlgorithm("SHA1");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA1_S128_B128_C10000));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha256Size128DecryptOpenSsl() {
        crypt.setHashAlgorithm("SHA256");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA256_S128_B128_C10000));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha256Size256DecryptOpenSsl() {
        crypt.setHashAlgorithm("SHA256");
        crypt.setKeySize(256);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA256_S256_B128_C10000));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha512Size128DecryptOpenSsl() {
        crypt.setHashAlgorithm("SHA512");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA512_S128_B128_C10000));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testSha512Size256DecryptOpenSsl() {
        crypt.setHashAlgorithm("SHA512");
        crypt.setKeySize(256);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(MdlConst.CRYPT_ITERATIONCOUNT);
        assertTrue(crypt.decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA512_S256_B128_C10000));
        assertEquals(MdlConst.CRYPT_TEST_PWD, crypt.getResult());
    }

    @Test
    public void testDeriveOpenSslKeyPbkdf2IterationZero() {
        byte[] baKey = MdlConst.CRYPT_TEST_KEY.getBytes(StandardCharsets.UTF_8);
        byte[] baSalt = new byte[] {1, 2, 3, 4, 5, 6, 7, 8};
        crypt.setHashAlgorithm("SHA512");
        crypt.setKeySize(MdlConst.CRYPT_KEYSIZE);
        crypt.setBlockSize(MdlConst.CRYPT_BLOCKSIZE);
        crypt.setIterationCount(0);
        assertTrue(crypt.deriveOpenSslKeyPbkdf2(baKey, baSalt));
        assertEquals(1, crypt.getIterationCount());
    }

    @Test
    public void testDeriveKeyPbkdf2() {
        byte[] baKey = MdlConst.CRYPT_TEST_KEY.getBytes(StandardCharsets.UTF_8);
        byte[] baSalt = new byte[] {1, 2, 3, 4, 5, 6, 7, 8};
        crypt.setHashAlgorithm("SHA256");
        crypt.setKeySize(256);
        crypt.setBlockSize(128);
        crypt.setIterationCount(1000);
        assertTrue(crypt.deriveKeyPbkdf2(baKey, baSalt));
    }
}

File: TestProject1/Class/UnitTest_ClsCrypt.cs
using CmnClsLib.Module;
using CmnClsLib.Class;
using CmnClsLib.Interface;
using System.Text;
using Xunit;
using Assert = Xunit.Assert;

namespace TestProject1.Class
{
    public class UnitTest_ClsCrypt
    {
        private ClsCrypt _crypt = new();

        // --------------------------------------------------------------------
        // public Boolean Encrypt(String strKey, String strPlain)
        // --------------------------------------------------------------------
        [Fact]
        public void MD5で文字列を暗号化および復号化できること()
        {
            _crypt.HashAlgorithm = "MD5";
            _crypt.KeySize = MdlConst.CRYPT_KEYSIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = 0;
            Assert.True(_crypt.Encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, _crypt.Result));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA1で文字列を暗号化および復号化できること()
        {
            _crypt.HashAlgorithm = "SHA1";
            _crypt.KeySize = MdlConst.CRYPT_KEYSIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, _crypt.Result));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA256サイズ128で文字列を暗号化および復号化できること()
        {
            _crypt.HashAlgorithm = "SHA256";
            _crypt.KeySize = MdlConst.CRYPT_KEYSIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, _crypt.Result));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA256サイズ256で文字列を暗号化および復号化できること()
        {
            _crypt.HashAlgorithm = "SHA256";
            _crypt.KeySize = 256;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, _crypt.Result));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA512サイズ128で文字列を暗号化および復号化できること()
        {
            _crypt.HashAlgorithm = "SHA512";
            _crypt.KeySize = MdlConst.CRYPT_KEYSIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, _crypt.Result));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA512サイズ256で文字列を暗号化および復号化できること()
        {
            _crypt.HashAlgorithm = "SHA512";
            _crypt.KeySize = 256;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Encrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.CRYPT_TEST_PWD));
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, _crypt.Result));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }

        // --------------------------------------------------------------------
        // public Boolean Decrypt(String strKey, String strPlain)
        // --------------------------------------------------------------------
        [Fact]
        public void MD5でOpenSSL互換暗号化文字列を復号化できること()
        {
            _crypt.HashAlgorithm = "MD5";
            _crypt.KeySize = MdlConst.CRYPT_KEYSIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = 0;
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_MD5_S128_B128_C0));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA1でOpenSSL互換暗号化文字列を復号化できること()
        {
            _crypt.HashAlgorithm = "SHA1";
            _crypt.KeySize = MdlConst.CRYPT_KEYSIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA1_S128_B128_C10000));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA256サイズ128でOpenSSL互換暗号化文字列を復号化できること()
        {
            _crypt.HashAlgorithm = "SHA256";
            _crypt.KeySize = MdlConst.CRYPT_KEYSIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA256_S128_B128_C10000));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA256サイズ256でOpenSSL互換暗号化文字列を復号化できること()
        {
            _crypt.HashAlgorithm = "SHA256";
            _crypt.KeySize = 256;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA256_S256_B128_C10000));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA512サイズ128でOpenSSL互換暗号化文字列を復号化できること()
        {
            _crypt.HashAlgorithm = "SHA512";
            _crypt.KeySize = ClsCrypt.DEFAULT_KEY_SIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA512_S128_B128_C10000));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }
        [Fact]
        public void SHA512サイズ256でOpenSSL互換暗号化文字列を復号化できること()
        {
            _crypt.HashAlgorithm = "SHA512";
            _crypt.KeySize = 256;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = MdlConst.CRYPT_ITERATIONCOUNT;
            Assert.True(_crypt.Decrypt(MdlConst.CRYPT_TEST_KEY, MdlConst.ENC_PASS_SHA512_S256_B128_C10000));
            Assert.Equal(MdlConst.CRYPT_TEST_PWD, _crypt.Result);
        }

        // --------------------------------------------------------------------
        // private Boolean DeriveOpenSslKeyPbkdf2(byte[] baKey, byte[] baSalt)
        // --------------------------------------------------------------------
        [Fact]
        public void DeriveOpenSslKeyPbkdf2でIterationCountが0の場合は1に変更されること()
        {
            byte[] baKey = Encoding.UTF8.GetBytes(MdlConst.CRYPT_TEST_KEY);
            byte[] baSalt = new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 };
            _crypt.HashAlgorithm = "SHA512";
            _crypt.KeySize = MdlConst.CRYPT_KEYSIZE;
            _crypt.BlockSize = MdlConst.CRYPT_BLOCKSIZE;
            _crypt.IterationCount = 0;
            Assert.True(_crypt.DeriveOpenSslKeyPbkdf2(baKey, baSalt));
            Assert.Equal(1, _crypt.IterationCount);
        }

        // --------------------------------------------------------------------
    }
}
